Frameworks/Foundation/NSURLProtocol_WinHTTP.mm

```diff
@@ -243,20 +244,35 @@ - (instancetype)initWithRequest:(NSURLRequest*)request
     };
 }
 
+// Helper function that executes a block on the client thread - used for client callbacks
+static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^callbackBlock)()) {
+    if (protocol->_clientThread) {
+        [protocol performSelector:@selector(_invokeBlock:) onThread:protocol->_clientThread withObject:callbackBlock waitUntilDone:NO];
+    } else {
+        callbackBlock();
+    }
+}
+
+// Helper function that just invokes a block
+- (void)_invokeBlock:(void (^)())block {
+    block();
+}
+
 - (void)startLoading {
     @synchronized(self) {
         if (_httpRequestOperation) {
             return;
         }
 
+        _clientThread = [NSThread currentThread];
```

Comment:
er, i'm assuming by 'before we finish here' you mean the whole set of callbacks and not just this call. pretty sure it's possible, though not in our internal use-cases. with our current impl of performSelector:, this raises an exception. I think this is acceptable since letting the client thread die seems like programmer error.